### ISSUE TITLE
Reorganize the modules for Function implementations

### ISSUE THREAD

**Issue by beam2d** [pfnet/chainer#331]
Thanks to many contributors, we now have many Function implementations under the `chainer.functions` module. It is convenient that user can access to them via `chainer.functions`, though the source code is less readable and searchable since they have no grouping based on the type of their functionalities. I want to reorganize the `chainer.functions` module. Note that I still want to keep the current interface that all functions can be accessed via `chainer.functions`.

**Comment by beam2d** [comment]
I made a possible structure below. The first level indicates the subpackage of chainer.functions (a.k.a. `F`), the second the module, and the third the functions. The functions listed below will also be imported to `F` for convenience (there are some exceptions). I will make a PR for this reorganization.

I also want to exclude create_huffman_tree from chainer.functions, since it is not a function on Variables. Possible options are 1.) moving it to a static method of BinaryHierarchicalSoftmax, or 2.) hiding it and making BinaryHierarchicalSoftmax directly accept word_counts.

- `connection`
  - `bilinear`
    - `Bilinear`
  - `convolution_2d`
    - `Convolution2D`
    - `convolution_2d`
  - `embed_id`
    - `EmbedID`
  - `inception`
    - `Inception`
  - `inception_bn`
    - `InceptionBN`
  - `linear`
    - `Linear`
    - `linear`
  - `parameter`
    - `Parameter`
- `math`
  - `basic_math`
    - `(Variable operators)`
  - `identity`
    - `identity`
  - `matmul`
    - `batch_matmul`
    - `matmul`
  - `sum`
    - `sum`
  - `trigonometric`
    - `sin`
    - `cos`
  - `exponential`
    - `exp`
    - `log`
- `activation`
  - `clipped_relu`
    - `clipped_relu`
  - `leaky_relu`
    - `leaky_relu`
  - `lstm`
    - `lstm`
  - `prelu`
    - `PReLU`
  - `relu`
    - `relu`
  - `sigmoid`
    - `sigmoid`
  - `softmax`
    - `softmax`
  - `softplus`
    - `softplus`
  - `tanh`
    - `tanh`
- `array`
  - `concat`
    - `concat`
  - `copy`
    - `copy`
  - `reshape`
    - `reshape`
  - `split_axis`
    - `split_axis`
- `loss`
  - `vae`
    - `bernoulli_nll`
    - `gaussian_kl_divergence`
    - `gaussian_nll`
  - `hierarchical_softmax`
    - `BinaryHierarchicalSoftmax`
  - `mean_squared_error`
    - `mean_squared_error`
  - `negative_sampling`
    - `NegativeSampling`
  - `sigmoid_cross_entropy`
    - `sigmoid_cross_entropy`
  - `softmax_cross_entropy`
    - `softmax_cross_entropy`
- `noize`
  - `dropout`
    - `dropout`
  - `gaussian`
    - `gaussian`
- `pooling`
  - `average_pooling_2d`
    - `average_pooling_2d`
  - `pooling_2d`
    - `Pooling2D` (not imported to `F`)
  - `max_pooling_2d`
    - `max_pooling_2d`
  - `spatial_pyramid_pooling_2d`
    - `spatial_pyramid_pooling_2d`
- `normalization`
  - `local_response_normalization`
    - `local_response_normalization`
  - `batch_normalization`
    - `BatchNormalization`
- `evaluation`
  - `accuracy`
    - `accuracy`

**Comment by beam2d** [comment]
Following an offline conversation with @unnonouno, I've decided to take the option 1 on the create_huffman_tree. Making BinaryHierarchicalSoftmax directly accept the word_counts would be discussed and done by another PR.

**Comment by beam2d** [comment]
I forgot the CaffeFunction. It will be remained as is (not imported to `F`).
